universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java

```diff
@@ -0,0 +1,78 @@
+package services.question;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.Locale;
+
+public class ReadOnlyQuestionServiceImpl implements ReadOnlyQuestionService {
+  private final ImmutableMap<String, ScalarType> scalars;
+  private final ImmutableMap<String, QuestionDefinition> questions;
+  private final ImmutableMap<String, QuestionDefinition> scalarParents;
+
+  private Locale preferredLocale = Locale.ENGLISH;
+
+  public ReadOnlyQuestionServiceImpl(ImmutableList<QuestionDefinition> questions) {
+    ImmutableMap.Builder<String, QuestionDefinition> questionMap = ImmutableMap.builder();
+    ImmutableMap.Builder<String, ScalarType> scalarMap = ImmutableMap.builder();
+    ImmutableMap.Builder<String, QuestionDefinition> scalarParentsMap = ImmutableMap.builder();
+    for (QuestionDefinition qd : questions) {
+      String questionPath = qd.getPath();
+      questionMap.put(questionPath, qd);
+      ImmutableMap<String, ScalarType> questionScalars = qd.getScalars();
+      questionScalars.entrySet().stream()
+          .forEach(
+              entry -> {
+                String fullPath = questionPath + '.' + entry.getKey();
+                scalarMap.put(fullPath, entry.getValue());
+                scalarParentsMap.put(fullPath, qd);
+              });
+    }
+    this.questions = questionMap.build();
+    this.scalars = scalarMap.build();
+    this.scalarParents = scalarParentsMap.build();
+  }
+
+  public ImmutableList<QuestionDefinition> getAllQuestions() {
+    return questions.values().asList();
+  }
+
+  public ImmutableMap<String, ScalarType> getAllScalars() {
+    return scalars;
+  }
+
+  public ImmutableMap<String, ScalarType> getPathScalars(String pathString)
+      throws InvalidPathException {
+    PathType pathType = this.getPathType(pathString);
+    if (pathType == PathType.QUESTION) {
+      return questions.get(pathString).getScalars(/* includeFullPath = */ true);
+    } else if (pathType == PathType.SCALAR) {
+      ScalarType scalarType = scalars.get(pathString);
+      return ImmutableMap.of(pathString, scalarType);
+    }
+    throw new InvalidPathException(pathString);
+  }
```

Comment:
Issue (blocking): do not use `==` for comparing enum types.

See https://engdoc.corp.google.com/eng/doc/devguide/java/practices/enums.md?cl=head#equals

Suggestion: use a `switch` here and L63